controllers/metal3remediation_controller.go

```diff
@@ -0,0 +1,219 @@
+	// This is for the future work, if finalizers are needed
+	//remediationMgr.SetFinalizer()
+
+	remediationType := remediationMgr.GetRemediationType()
+
+	if remediationType == capm3.RebootRemediationStrategy {
+		// If no phase set, default to pending
+		if remediationMgr.GetRemediationPhase() == "" {
+			remediationMgr.SetRemediationPhase(capm3.PhaseRunning)
+		}
+
+		switch remediationMgr.GetRemediationPhase() {
+		case capm3.PhaseRunning:
+			// host is not rebooted yet
+			if remediationMgr.GetLastRemediatedTime() == nil {
+				r.Log.Info("Rebooting the host")
+				err := remediationMgr.SetAnnotation(ctx, "reboot")
+				if err != nil {
+					return ctrl.Result{}, errors.Wrap(err, "error setting reboot annotation")
+				}
+				now := metav1.Now()
+				remediationMgr.SetLastRemediationTime(&now)
+				remediationMgr.IncreaseRetryCount()
+			}
+
+			if remediationMgr.RetryLimitIsSet() && !remediationMgr.HasReachRetryLimit() {
+				okToRemediate, nextRemediation := remediationMgr.TimeToRemediate(remediationMgr.GetTimeout().Duration)
+
+				if okToRemediate {
+					err := remediationMgr.SetAnnotation(ctx, "reboot")
+					if err != nil {
+						return ctrl.Result{}, errors.Wrapf(err, "error setting reboot annotation")
+					}
+					now := metav1.Now()
+					remediationMgr.SetLastRemediationTime(&now)
+					remediationMgr.IncreaseRetryCount()
+				}
+
+				if nextRemediation > 0 {
+					// Not yet time to remediatate, requeue
+					return ctrl.Result{RequeueAfter: nextRemediation}, nil
+				}
+
+			} else {
+				remediationMgr.SetRemediationPhase(capm3.PhaseWaiting)
+			}
+		case capm3.PhaseWaiting:
+			okToStop, nextCheck := remediationMgr.TimeToRemediate(remediationMgr.GetTimeout().Duration)
+
+			if okToStop {
+				remediationMgr.SetRemediationPhase(capm3.PhaseDeleting)
+				// When machine is still unhealthy after remediation, setting of OwnerRemediatedCondition
+				// moves control to CAPI machine controller. The owning controller will do
+				// preflight checks and handles the Machine deletion
+				err = remediationMgr.SetOwnerRemediatedCondition(ctx)
+				if err != nil {
+					return ctrl.Result{}, errors.Wrapf(err, "error setting condition")
+				}
+
+				// Remediation failed set unhealthy annotation on BMH
+				// This prevents BMH to be selected as a host.
+				err = remediationMgr.SetAnnotation(ctx, "unhealthy")
+				if err != nil {
+					return ctrl.Result{}, errors.Wrapf(err, "error setting unhealthy annotation")
+				}
+			}
+
+			if nextCheck > 0 {
+				// Not yet time to stop remediation, requeue
+				return ctrl.Result{RequeueAfter: nextCheck}, nil
+			}
+		default:
+
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *Metal3RemediationReconciler) reconcileDelete(ctx context.Context,
+	remediationMgr baremetal.RemediationManagerInterface,
+) (ctrl.Result, error) {
+
+	// metal3remediation is marked for deletion and ready to be deleted,
+	// so remove the finalizer.
+	// remediationMgr.UnsetFinalizer()
```

Comment:
Currently we don't use finalizers but the functions are implemented. This is for the future work but, maybe it is better to remove it here.

Reply:
Yes, I think it is better if we remove it, if there is no use of it.

Reply:
WDYT about removing the `unhealthy` annotation from the host once the remediation request is being deleted?

Reply:
Do we really want to remove the `unhealthy` annotation? If our host is not healthy and possible needs manual maintenance annotation will make sure the host is not getting provisioned before annotation is removed manually.

Reply:
if `metal3remediation` was deleted, it means that the host is healthy again, isn't it?

Reply:
That is true in cases where MHC find out Machine healthy again and deletes the metal3remediation. 

But when RC tryies to reboot host n times and timeouts after last retry, it will mark BMH as unhealthy and set condition on Machine for deletion. So in latter case we are not sure if host is healthy or not and is it capable of serving as a host for any new Machine objects.

Reply:
Good point.
So before removing the finalizer we can check the `metal3remediation` CR.
If remediation is failed and machine is going to be deleted we won't remove the unhealthy annotation.
Does that make sense?